File: scripts/extract-prompts.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
智能提示词数据提取脚本 v3
- 完整提取提示词内容
- 智能识别 AI 模型（从内容中检测）
- 自动判断难度等级
- 提取有效标签（中英文关键词）
- 过滤空内容和占位符
"""

import os
import re
import json
import yaml
from pathlib import Path
from typing import Dict, List, Any, Optional

# 模型识别规则（优先级从高到低）
MODEL_PATTERNS = {
    'Gemini': [r'gemini', r'google\s*ai', r'nano\s*banana'],
    'Grok': [r'grok', r'xAI'],
    'Seedream': [r'seedream', r'seed\s*ream'],
    'DALL-E': [r'dall-e', r'dalle', r'openai'],
    'Stable Diffusion': [r'stable\s*diffusion', r'sd\s*xl', r'sdxl', r'comfyui'],
    'Flux': [r'flux', r'black\s*forest'],
    'Ideogram': [r'ideogram'],
    'Leonardo': [r'leonardo'],
    'Adobe Firefly': [r'firefly', r'adobe'],
    'Midjourney': [r'midjourney', r'mj\s*v?\d', r'niji\s*\d', r'/imagine'],
}

# 难度判断规则
DIFFICULTY_KEYWORDS = {
    'advanced': [
        r'参数', r'parameter', r'权重', r'weight', r'cfg',
        r'controlnet', r'ip-adapter', r'workflow', r'复杂',
        r'advanced', r'高级', r'专业', r'expert', r'--ar',
        r'--v', r'--s', r'--c', r'--q', r'--style',
        r'公式', r'formula', r'技巧', r'technique'
    ],
    'intermediate': [
        r'提示词', r'prompt', r'风格', r'style',
        r'方法', r'method', r'教程', r'tutorial',
        r'步骤', r'step'
    ],
    'beginner': [
        r'简单', r'simple', r'基础', r'basic', r'入门',
        r'beginner', r'新手', r'快速', r'quick', r'一键'
    ]
}

# 标签提取关键词（中英文）
TAG_KEYWORDS = {
    '3D渲染': [r'3d', r'3D', r'c4d', r'octane', r'render', r'渲染', r'三维'],
    '国风': [r'国风', r'中国风', r'东方', r'chinese', r'oriental', r'汉服', r'hanfu', r'古风'],
    '赛博朋克': [r'赛博朋克', r'cyberpunk', r'cyber'],
    '写实': [r'写实', r'realistic', r'photorealistic', r'真实', r'照片级'],
    '动漫': [r'动漫', r'anime', r'manga', r'二次元', r'卡通'],
    '油画': [r'油画', r'oil\s*painting', r'classical', r'古典'],
    '摄影': [r'摄影', r'photography', r'photo', r'镜头', r'camera'],
    '抽象': [r'抽象', r'abstract'],
    '人物': [r'人物', r'portrait', r'character', r'角色', r'人像'],
    '风景': [r'风景', r'landscape', r'scenery', r'自然'],
    '建筑': [r'建筑', r'architecture', r'building'],
    '产品': [r'产品', r'product', r'商品', r'商业'],
    '概念艺术': [r'概念', r'concept', r'概念艺术'],
    '科幻': [r'科幻', r'sci-fi', r'science\s*fiction', r'未来', r'futuristic'],
    '奇幻': [r'奇幻', r'fantasy', r'魔幻', r'magical'],
    '恐怖': [r'恐怖', r'horror', r'惊悚'],
    '可爱': [r'可爱', r'cute', r'kawaii', r'萌', r'卡通'],
    '复古': [r'复古', r'retro', r'vintage', r'怀旧'],
    '极简': [r'极简', r'minimal', r'minimalist', r'简约'],
    '超现实': [r'超现实', r'surreal', r'surrealism'],
    '蒸汽朋克': [r'蒸汽朋克', r'steampunk'],
    'AI艺术': [r'ai\s*art', r'ai\s*艺术', r'人工智能'],
    '电影感': [r'电影', r'cinematic', r'cinema', r'film'],
    '分焦镜头': [r'分焦', r'split\s*diopter', r'双焦'],
}


def detect_model(content: str, frontmatter_model: str = '') -> str:
    """从内容中智能识别 AI 模型"""
    content_lower = content.lower()
    
    # 先检查 frontmatter 中的模型
    if frontmatter_model:
        for model, patterns in MODEL_PATTERNS.items():
            for pattern in patterns:
                if re.search(pattern, frontmatter_model.lower(), re.IGNORECASE):
                    return model
        # 如果 frontmatter 有值但没匹配到，直接使用
        return frontmatter_model
    
    # 从内容中检测（按优先级）
    for model, patterns in MODEL_PATTERNS.items():
        for pattern in patterns:
            if re.search(pattern, content_lower, re.IGNORECASE):
                return model
    
    return 'Unknown'


def detect_difficulty(content: str) -> str:
    """根据内容复杂度判断难度"""
    content_lower = content.lower()
    
    # 统计各难度关键词出现次数
    scores = {'beginner': 0, 'intermediate': 0, 'advanced': 0}
    
    for difficulty, keywords in DIFFICULTY_KEYWORDS.items():
        for keyword in keywords:
            matches = re.findall(keyword, content_lower, re.IGNORECASE)
            scores[difficulty] += len(matches)
    
    # 根据内容长度调整
    if len(content) > 1000:
        scores['advanced'] += 2
    elif len(content) > 500:
        scores['intermediate'] += 2
    
    # 返回得分最高的难度
    max_difficulty = max(scores, key=lambda k: scores[k])
    
    # 如果得分都很低，默认为 intermediate
    if scores[max_difficulty] == 0:
        return 'intermediate'
    
    return max_difficulty


def extract_tags(content: str, existing_tags: List[str]) -> List[str]:
    """提取有效标签，去除占位符"""
    # 过滤掉"待处理"等占位符
    valid_tags = [tag for tag in existing_tags if tag not in ['待处理', 'todo', 'placeholder']]
    
    # 从内容中提取标签
    content_lower = content.lower()
    extracted_tags = []
    
    for tag, keywords in TAG_KEYWORDS.items():
        for keyword in keywords:
            if re.search(keyword, content_lower, re.IGNORECASE):
                if tag not in valid_tags and tag not in extracted_tags:
                    extracted_tags.append(tag)
                break
    
    # 合并标签，最多保留 5 个
    all_tags = valid_tags + extracted_tags
    return list(dict.fromkeys(all_tags))[:5]  # 去重并限制数量


def clean_prompt(content: str) -> str:
    """清理提示词内容，提取核心部分"""
    if not content or content.strip() in ['Prompt:', '提示词:', '']:
        return ''
    
    # 移除常见的多余文字模式
    patterns_to_remove = [
        r'^(Prompt:|提示词:)\s*',
        r'(复制即用|一键出图|咒语)[：:]\s*',
        r'/imagine\s+prompt:\s*',
    ]
    
    cleaned = content.strip()
    
    for pattern in patterns_to_remove:
        cleaned = re.sub(pattern, '', cleaned, flags=re.IGNORECASE | re.MULTILINE)
    
    return cleaned.strip()


def parse_markdown_file(file_path: Path) -> Optional[Dict[str, Any]]:
    """解析单个 markdown 文件"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        # 分离 frontmatter 和正文
        if not content.startswith('---'):
            return None
        
        parts = content.split('---', 2)
        if len(parts) < 3:
            return None
        
        frontmatter_str = parts[1]
        body = parts[2]
        
        # 解析 frontmatter
        try:
            frontmatter = yaml.safe_load(frontmatter_str)
        except yaml.YAMLError:
            return None
        
        # 提取各个部分
        sections = {}
        current_section = None
        current_content = []
        
        for line in body.split('\n'):
            if line.startswith('## '):
                if current_section:
                    sections[current_section] = '\n'.join(current_content).strip()
                current_section = line[3:].strip()
                current_content = []
            else:
                current_content.append(line)
        
        if current_section:
            sections[current_section] = '\n'.join(current_content).strip()
        
        # 提取提示词
        prompt_text = sections.get('Prompt', '')
        negative_prompt = sections.get('Negative Prompt', '')
        
        # 清理提示词
        cleaned_prompt = clean_prompt(prompt_text)
        
        # 如果清理后为空，跳过这个文件
        if not cleaned_prompt:
            return None
        
        # 智能识别模型
        detected_model = detect_model(body, frontmatter.get('model', ''))
        
        # 判断难度
        detected_difficulty = detect_difficulty(body)
        
        # 提取标签
        existing_tags = frontmatter.get('tags', [])
        if isinstance(existing_tags, str):
            existing_tags = [existing_tags]
        extracted_tags = extract_tags(body, existing_tags)
        
        # 提取参数
        parameters = {}
        if 'Parameters' in sections:
            param_text = sections['Parameters']
            for line in param_text.split('\n'):
                if '|' in line and not line.strip().startswith('|---'):
                    parts = [p.strip() for p in line.split('|') if p.strip()]
                    if len(parts) == 2 and parts[0] not in ['Setting', '参数']:
                        parameters[parts[0]] = parts[1]
        
        # 构建数据结构
        result = {
            'title': frontmatter.get('title', ''),
            'slug': frontmatter.get('slug', ''),
            'model': detected_model,
            'category': frontmatter.get('category', 'uncategorized'),
            'tags': extracted_tags,
            'difficulty': detected_difficulty,
            'cover': frontmatter.get('cover', ''),
            'date': str(frontmatter.get('date', '')),
            'source': frontmatter.get('source', ''),
            'author': frontmatter.get('author', 'Unknown'),
            'prompt': cleaned_prompt,
            'negativePrompt': negative_prompt if negative_prompt and negative_prompt != '(none provided)' else '',
            'parameters': parameters
        }
        
        return result
    except Exception as e:
        print(f"❌ 处理失败 {file_path.name}: {e}")
        return None


def main():
    """主函数"""
    content_dir = Path('/Users/mac/.hermes/profiles/cgfan/workspace/cgfan-web/content/prompts')
    output_file = Path('/Users/mac/.hermes/profiles/cgfan/workspace/cgfan-web/lib/prompts-data.json')
    
    all_prompts = []
    error_count = 0
    skipped_count = 0
    
    # 遍历所有 markdown 文件
    for md_file in content_dir.rglob('*.md'):
        prompt_data = parse_markdown_file(md_file)
        if prompt_data:
            all_prompts.append(prompt_data)
        else:
            skipped_count += 1
    
    # 按日期排序（最新的在前）
    all_prompts.sort(key=lambda x: x.get('date', ''), reverse=True)
    
    # 写入 JSON
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(all_prompts, f, ensure_ascii=False, indent=2)
    
    # 统计信息
    model_stats = {}
    difficulty_stats = {}
    
    for prompt in all_prompts:
        model = prompt.get('model', 'Unknown')
        difficulty = prompt.get('difficulty', 'Unknown')
        
        model_stats[model] = model_stats.get(model, 0) + 1
        difficulty_stats[difficulty] = difficulty_stats.get(difficulty, 0) + 1
    
    print("\n" + "="*60)
    print("✅ 数据提取完成！")
    print("="*60)
    print(f"📊 总计: {len(all_prompts)} 条提示词")
    print(f"⚠️  跳过: {skipped_count} 个文件（空内容或格式错误）")
    print("\n🤖 模型分布:")
    for model, count in sorted(model_stats.items(), key=lambda x: x[1], reverse=True):
        print(f"   {model}: {count}")
    print("\n📈 难度分布:")
    for difficulty, count in sorted(difficulty_stats.items(), key=lambda x: x[1], reverse=True):
        print(f"   {difficulty}: {count}")
    print("="*60)


if __name__ == '__main__':
    main()

File: scripts/extract-prompts.ts
#!/usr/bin/env tsx
import fs from 'fs';
import path from 'path';
import matter from 'gray-matter';

// 模型识别规则（优先级从高到低）
const MODEL_PATTERNS: Record<string, RegExp[]> = {
  'Gemini': [/gemini/i, /google\s*ai/i, /nano\s*banana/i],
  'Grok': [/grok/i, /xAI/i],
  'Seedream': [/seedream/i, /seed\s*ream/i],
  'DALL-E': [/dall-e/i, /dalle/i, /openai/i],
  'Stable Diffusion': [/stable\s*diffusion/i, /sd\s*xl/i, /sdxl/i, /comfyui/i],
  'Flux': [/flux/i, /black\s*forest/i],
  'Ideogram': [/ideogram/i],
  'Leonardo': [/leonardo/i],
  'Adobe Firefly': [/firefly/i, /adobe/i],
  'Midjourney': [/midjourney/i, /mj\s*v?\d/i, /niji\s*\d/i, /\/imagine/i],
  'GPT-image2': [/gpt[\s-]*image[\s-]*2/i, /gpt[\s-]*image2/i, /chatgpt/i, /gpt-4/i, /gpt4/i],
};

// 难度判断规则
const DIFFICULTY_KEYWORDS = {
  advanced: [
    /参数/i, /parameter/i, /权重/i, /weight/i, /cfg/i,
    /controlnet/i, /ip-adapter/i, /workflow/i, /复杂/i,
    /advanced/i, /高级/i, /专业/i, /expert/i, /--ar/i,
    /--v/i, /--s/i, /--c/i, /--q/i, /--style/i,
    /公式/i, /formula/i, /技巧/i, /technique/i
  ],
  intermediate: [
    /提示词/i, /prompt/i, /风格/i, /style/i,
    /方法/i, /method/i, /教程/i, /tutorial/i,
    /步骤/i, /step/i
  ],
  beginner: [
    /简单/i, /simple/i, /基础/i, /basic/i, /入门/i,
    /beginner/i, /新手/i, /快速/i, /quick/i, /一键/i
  ]
};

// 标签提取关键词（中英文）
const TAG_KEYWORDS: Record<string, RegExp[]> = {
  '3D渲染': [/3d/i, /c4d/i, /octane/i, /render/i, /渲染/i, /三维/i],
  '国风': [/国风/i, /中国风/i, /东方/i, /chinese/i, /oriental/i, /汉服/i, /hanfu/i, /古风/i],
  '赛博朋克': [/赛博朋克/i, /cyberpunk/i, /cyber/i],
  '写实': [/写实/i, /realistic/i, /photorealistic/i, /真实/i, /照片级/i],
  '动漫': [/动漫/i, /anime/i, /manga/i, /二次元/i, /卡通/i],
  '油画': [/油画/i, /oil\s*painting/i, /classical/i, /古典/i],
  '摄影': [/摄影/i, /photography/i, /photo/i, /镜头/i, /camera/i],
  '抽象': [/抽象/i, /abstract/i],
  '人物': [/人物/i, /portrait/i, /character/i, /角色/i, /人像/i],
  '风景': [/风景/i, /landscape/i, /scenery/i, /自然/i],
  '建筑': [/建筑/i, /architecture/i, /building/i],
  '产品': [/产品/i, /product/i, /商品/i, /商业/i],
  '概念艺术': [/概念/i, /concept/i, /概念艺术/i],
  '科幻': [/科幻/i, /sci-fi/i, /science\s*fiction/i, /未来/i, /futuristic/i],
  '奇幻': [/奇幻/i, /fantasy/i, /魔幻/i, /magical/i],
  '恐怖': [/恐怖/i, /horror/i, /惊悚/i],
  '可爱': [/可爱/i, /cute/i, /kawaii/i, /萌/i, /卡通/i],
  '复古': [/复古/i, /retro/i, /vintage/i, /怀旧/i],
  '极简': [/极简/i, /minimal/i, /minimalist/i, /简约/i],
  '超现实': [/超现实/i, /surreal/i, /surrealism/i],
  '蒸汽朋克': [/蒸汽朋克/i, /steampunk/i],
  'AI艺术': [/ai\s*art/i, /ai\s*艺术/i, /人工智能/i],
  '电影感': [/电影/i, /cinematic/i, /cinema/i, /film/i],
  '分焦镜头': [/分焦/i, /split\s*diopter/i, /双焦/i],
};

// 需要移除的占位符标签
const PLACEHOLDER_TAGS = ['待处理', 'todo', 'placeholder', 'TODO', '待整理'];

function detectModel(content: string, frontmatterModel: string = ''): string {
  const contentLower = content.toLowerCase();
  
  // 优先从正文内容中检测（按优先级）
  for (const [model, patterns] of Object.entries(MODEL_PATTERNS)) {
    for (const pattern of patterns) {
      if (pattern.test(contentLower)) {
        return model;
      }
    }
  }
  
  // 正文没检测到，才用 frontmatter
  if (frontmatterModel) {
    for (const [model, patterns] of Object.entries(MODEL_PATTERNS)) {
      for (const pattern of patterns) {
        if (pattern.test(frontmatterModel)) {
          return model;
        }
      }
    }
    return frontmatterModel;
  }
  
  return 'Common';
}

function detectDifficulty(content: string): string {
  const contentLower = content.toLowerCase();
  const scores: Record<string, number> = { beginner: 0, intermediate: 0, advanced: 0 };
  
  for (const [difficulty, keywords] of Object.entries(DIFFICULTY_KEYWORDS)) {
    for (const keyword of keywords) {
      const matches = contentLower.match(new RegExp(keyword, 'gi'));
      if (matches) {
        scores[difficulty] += matches.length;
      }
    }
  }
  
  // 根据内容长度调整难度
  const len = content.length;
  if (len < 200) {
    scores.beginner += 3;      // 短提示词 → 入门
  } else if (len < 600) {
    scores.intermediate += 2;  // 中等长度 → 进阶
  } else if (len > 1200) {
    scores.advanced += 3;      // 超长提示词 → 高级
  } else if (len > 600) {
    scores.intermediate += 1;  // 较长 → 进阶
  }
  
  // 返回得分最高的难度
  const maxDifficulty = Object.keys(scores).reduce((a, b) => scores[a] > scores[b] ? a : b);
  
  // 如果得分都很低，根据内容长度决定
  if (scores[maxDifficulty] === 0) {
    return len < 300 ? 'beginner' : len > 1000 ? 'advanced' : 'intermediate';
  }
  
  return maxDifficulty;
}

function extractTags(content: string, existingTags: string[]): string[] {
  // 过滤掉占位符标签
  const validTags = existingTags.filter(tag => !PLACEHOLDER_TAGS.includes(tag));
  
  // 从内容中提取标签
  const contentLower = content.toLowerCase();
  const extractedTags: string[] = [];
  
  for (const [tag, keywords] of Object.entries(TAG_KEYWORDS)) {
    for (const keyword of keywords) {
      if (keyword.test(contentLower)) {
        if (!validTags.includes(tag) && !extractedTags.includes(tag)) {
          extractedTags.push(tag);
        }
        break;
      }
    }
  }
  
  // 合并标签，最多保留 5 个
  const allTags = [...validTags, ...extractedTags];
  return Array.from(new Set(allTags)).slice(0, 5);
}

function cleanPrompt(content: string): string {
  if (!content || ['Prompt:', '提示词:', ''].includes(content.trim())) {
    return '';
  }
  
  // 移除常见的多余文字模式
  const patternsToRemove = [
    /^(Prompt:|提示词:)\s*/i,
    /(复制即用|一键出图|咒语)[：:]\s*/i,
    /\/imagine\s+prompt:\s*/i,
  ];
  
  let cleaned = content.trim();
  
  for (const pattern of patternsToRemove) {
    cleaned = cleaned.replace(pattern, '');
  }
  
  return cleaned.trim();
}

function normalizeDate(dateValue: any): string {
  if (!dateValue) return '';
  
  // 如果是 Date 对象，检查是否有非零时间分量（表示是精确时间戳）
  if (dateValue instanceof Date && !isNaN(dateValue.getTime())) {
    const hours = dateValue.getUTCHours();
    const minutes = dateValue.getUTCMinutes();
    const seconds = dateValue.getUTCSeconds();
    // 如果有时分秒分量（非全部为0），说明原本是 ISO 时间戳，保留完整 ISO
    if (hours !== 0 || minutes !== 0 || seconds !== 0) {
      // 使用 +08:00 时区输出，避免 toISOString() 转 UTC 导致排序混乱
      const pad = (n: number) => String(n).padStart(2, '0');
      const y = dateValue.getFullYear();
      const m = pad(dateValue.getMonth() + 1);
      const d = pad(dateValue.getDate());
      const hh = pad(dateValue.getHours());
      const mm = pad(dateValue.getMinutes());
      const ss = pad(dateValue.getSeconds());
      return `${y}-${m}-${d}T${hh}:${mm}:${ss}+08:00`;
    }
    // 全部为0，说明原本只是日期，只输出 YYYY-MM-DD
    return dateValue.toISOString().split('T')[0];
  }
  
  // 如果是 ISO 8601 时间戳字符串（含 T 和时区），直接保留
  if (typeof dateValue === 'string' && /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}/.test(dateValue)) {
    return dateValue;
  }
  
  // 如果已经是 YYYY-MM-DD 格式
  if (typeof dateValue === 'string' && /^\d{4}-\d{2}-\d{2}$/.test(dateValue)) {
    return dateValue;
  }
  
  // 如果是字符串但可以解析为日期
  try {
    const date = new Date(dateValue);
    if (!isNaN(date.getTime())) {
      return date.toISOString().split('T')[0];
    }
  } catch {
    // 忽略错误
  }
  
  // 如果无法解析，返回原值
  return String(dateValue);
}

interface PromptData {
  title: string;
  slug: string;
  model: string;
  category: string;
  tags: string[];
  difficulty: string;
  cover: string;
  images: string[];
  date: string;
  added: string; // 上传日期，用于排序
  source: string;
  sourceLink: string;
  author: string;
  authorLink?: string;
  prompt: string;
  negativePrompt: string;
  parameters: Record<string, string>;
  rating?: {
    composition: number;
    color: number;
    lighting: number;
    detail: number;
    creativity: number;
    technical: number;
    aesthetic: number;
    curation: number;
    total: number;
  };
}

function parseMarkdownFile(filePath: string): PromptData | null {
  try {
    const content = fs.readFileSync(filePath, 'utf-8');
    const { data: frontmatter, content: body } = matter(content);
    
    // 提取各个部分
    const sections: Record<string, string> = {};
    let currentSection = '';
    let currentContent: string[] = [];
    
    for (const line of body.split('\n')) {
      if (line.startsWith('## ')) {
        if (currentSection) {
          sections[currentSection] = currentContent.join('\n').trim();
        }
        currentSection = line.slice(3).trim();
        currentContent = [];
      } else {
        currentContent.push(line);
      }
    }
    
    if (currentSection) {
      sections[currentSection] = currentContent.join('\n').trim();
    }
    
    // 提取提示词 - 优先使用 ## Prompt 章节，否则使用整个正文
    let promptText = sections['Prompt'] || '';
    const negativePrompt = sections['Negative Prompt'] || '';
    
    // 如果没有 ## Prompt 章节，使用整个正文作为提示词
    if (!promptText) {
      // 检查是否有类似 "## Prompt 1:" 的章节
      const promptSections = Object.keys(sections).filter(k => k.toLowerCase().startsWith('prompt'));
      if (promptSections.length > 0) {
        promptText = promptSections.map(k => sections[k]).join('\n\n');
      } else {
        promptText = body.trim();
      }
    }
    
    // 清理提示词
    const cleanedPrompt = cleanPrompt(promptText);
    
    // 如果清理后为空或太短（少于50字符），跳过这个文件
    if (!cleanedPrompt || cleanedPrompt.length < 50) {
      return null;
    }
    
    // 智能识别模型
    const detectedModel = detectModel(body, frontmatter.model || '');
    
    // 判断难度
    const detectedDifficulty = detectDifficulty(body);
    
    // 提取标签
    const existingTags = Array.isArray(frontmatter.tags) ? frontmatter.tags : [frontmatter.tags || ''];
    const extractedTags = extractTags(body, existingTags);
    
    // 提取参数
    const parameters: Record<string, string> = {};
    if (sections['Parameters']) {
      const paramText = sections['Parameters'];
      for (const line of paramText.split('\n')) {
        if (line.includes('|') && !line.trim().startsWith('|---')) {
          const parts = line.split('|').map(p => p.trim()).filter(p => p);
          if (parts.length === 2 && !['Setting', '参数'].includes(parts[0])) {
            parameters[parts[0]] = parts[1];
          }
        }
      }
    }
    
    // 读取评分字段
    const rating = frontmatter.rating ? {
      composition: Number(frontmatter.rating.composition) || 0,
      color: Number(frontmatter.rating.color) || 0,
      lighting: Number(frontmatter.rating.lighting) || 0,
      detail: Number(frontmatter.rating.detail) || 0,
      creativity: Number(frontmatter.rating.creativity) || 0,
      technical: Number(frontmatter.rating.technical) || 0,
      aesthetic: Number(frontmatter.rating.aesthetic) || 0,
      curation: Number(frontmatter.rating.curation) || 0,
      total: Number(frontmatter.rating.total) || 0
    } : undefined;

    return {
      title: frontmatter.title || '',
      slug: frontmatter.slug || path.basename(filePath, '.md'), // fallback 用文件名
      model: detectedModel,
      category: frontmatter.category || 'uncategorized',
      tags: extractedTags,
      difficulty: detectedDifficulty,
      cover: frontmatter.cover || '',
      images: Array.isArray(frontmatter.images) ? frontmatter.images : (frontmatter.cover ? [frontmatter.cover] : []),
      date: String(frontmatter.date || ''),
      added: normalizeDate(frontmatter.added || frontmatter.date || ''), // 保留原始时间戳或纯日期
      source: frontmatter.source || '',
      sourceLink: frontmatter.sourceLink || (frontmatter.source && frontmatter.source.startsWith('http') ? frontmatter.source : ''),
      author: frontmatter.author || 'Unknown',
      authorLink: frontmatter.authorLink || '',
      prompt: cleanedPrompt,
      negativePrompt: negativePrompt && negativePrompt !== '(none provided)' ? negativePrompt : '',
      parameters,
      rating
    };
  } catch (error) {
    console.error(`❌ 处理失败 ${path.basename(filePath)}:`, error);
    return null;
  }
}

async function main() {
  const contentDir = path.join(process.cwd(), 'content/prompts');
  
  const allPrompts: PromptData[] = [];
  let skippedCount = 0;
  
  // 遍历所有 markdown 文件
  function walkDir(dir: string) {
    const files = fs.readdirSync(dir);
    for (const file of files) {
      const filePath = path.join(dir, file);
      const stat = fs.statSync(filePath);
      if (stat.isDirectory()) {
        walkDir(filePath);
      } else if (file.endsWith('.md')) {
        const promptData = parseMarkdownFile(filePath);
        if (promptData) {
          // 添加文件修改时间用于排序
          (promptData as any).mtime = stat.mtimeMs;
          // 如果 added 是纯日期（无时间分量），用固定正午时间补全，保留日期语义
          if (promptData.added && /^\d{4}-\d{2}-\d{2}$/.test(promptData.added)) {
            promptData.added = `${promptData.added}T12:00:00.000+08:00`;
          }
          allPrompts.push(promptData);
        } else {
          skippedCount++;
        }
      }
    }
  }
  
  walkDir(contentDir);
  
  // 按上传时间升序排列（最早上传在前，最新上传在后）
  // 优先使用 added 字段（含 ISO 时间戳），否则 fallback 到 mtime
  allPrompts.sort((a, b) => {
    const addedA = a.added || '';
    const addedB = b.added || '';
    
    // 如果 added 含 ISO 时间戳（带 T），精确排序
    const hasTimeA = addedA.includes('T');
    const hasTimeB = addedB.includes('T');
    
    if (hasTimeA && hasTimeB) {
      // 都有 ISO 时间戳 → 按时间排序
      return addedA.localeCompare(addedB);
    } else if (hasTimeA && !hasTimeB) {
      // A 有精确时间 → A 排在 B 后面（B 是同一天，先显示）
      return 1;
    } else if (!hasTimeA && hasTimeB) {
      // B 有精确时间 → B 排在 A 后面
      return -1;
    } else {
      // 都没有精确时间 → 用 mtime 排序
      const mtimeA = (a as any).mtime || 0;
      const mtimeB = (b as any).mtime || 0;
      return mtimeA - mtimeB;
    }
  });
  
  // 同步到 Supabase
  console.log('🔄 正在同步到 Supabase...');
  const { upsertPrompts } = await import('./supabase-utils');
  const upsertedCount = await upsertPrompts(allPrompts);
  
  console.log(`✅ 提取完成：${allPrompts.length} 条提示词已同步到 Supabase（成功 ${upsertedCount} 条）`);
  
  // 统计信息
  const modelStats: Record<string, number> = {};
  const difficultyStats: Record<string, number> = {};
  
  for (const prompt of allPrompts) {
    modelStats[prompt.model] = (modelStats[prompt.model] || 0) + 1;
    difficultyStats[prompt.difficulty] = (difficultyStats[prompt.difficulty] || 0) + 1;
  }
  
  console.log('\n' + '='.repeat(60));
  console.log('✅ 数据提取完成！');
  console.log('='.repeat(60));
  console.log(`📊 总计: ${allPrompts.length} 条提示词`);
  console.log(`⚠️  跳过: ${skippedCount} 个文件（空内容或格式错误）`);
  console.log('\n🤖 模型分布:');
  Object.entries(modelStats)
    .sort((a, b) => b[1] - a[1])
    .forEach(([model, count]) => console.log(`   ${model}: ${count}`));
  console.log('\n📈 难度分布:');
  Object.entries(difficultyStats)
    .sort((a, b) => b[1] - a[1])
    .forEach(([difficulty, count]) => console.log(`   ${difficulty}: ${count}`));
  console.log('='.repeat(60));
}

main();
